### ISSUE TITLE
Are select queries aware of detach? 

### ISSUE THREAD

**Issue by gauravphoenix** [ClickHouse/ClickHouse#31612]
If my select query is running and I issue a detach command on a partition(which is being processed by the select query in question), will the select query still read the partition data? 

I am not able to test this because my select queries are super fast so I would like to confirm it.

**Comment by gauravphoenix** [comment]
I was able to experiment with this by putting a lot of data in my table and then performing a select. 
Not only I did `detach`, I went to the store directory and did `rm -rf *` to remove the detached partitions. 

so my question is no longer about "if" but rather "how". 

when select queries are being run, and I have already performed detach and removed the files, where is the data being read from?

**Comment by gauravphoenix** [comment]
The mystery is solved now, I did some more investigations and it seems that when `detach` command is executed, the data is **not** removed immediately from the store directory when `select` queries are running. The copy to detach directory still occurs but not the removal and hence select queries are able to see the data. 


it will be nice to have documentation updated to reflect this

**Comment by alexey-milovidov** [comment]
Yes, the data parts are refcounted, and removed only after the latest query finished.

**Comment by gauravphoenix** [comment]
@alexey-milovidov does refcount apply to attach commands too? Can I safely run attach commands when select queries are being run?

**Comment by alexey-milovidov** [comment]
Yes, you can.

The same mechanics allows to run multiple INSERT and SELECT queries concurrently and multiple background merges without locks. This is fundamental for ClickHouse.

**Comment by alexey-milovidov** [comment]
And even more... if you use Atomic database (by default), you can DROP table during SELECT without any locks and waiting :)
It's magic.
